### DISCUSSION TITLE
namespace qualifiers::

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1165] [Q&A] [answered]
Just to check whether my understanding is totally wrong--
So when a project contains several or dozens of modules,a lot of structs,constants,functions and methods inside these modules,now when making use of all these stuffs,we have to use namespace qualifiers on almost every of the aforementioned like below:

AAA::consts c;
BBB::enums e;
fn void CCC::functon(AAA:arg1,BBB::arg2,...){}
struct DDD
{
    AAA::T t;
   BBB::U u;
...
}
Is this mandatory or there is just option ways leave to user to just write like fn void function(arg1,arg2,...) ?

Thanks and best regards,
Peter

**Comment by lerno** [comment]
No that's not correct. You are required to use the shortened namespace on imported constants, globals, functions and macros. When I say "shortened namespace", I mean that if you have something like `my_function` in `std::io::os`, you only need to use `os::my_function` as long as that is unambiguous in the context. That is, you're not importing some other namespace like `my::proj::os` which has the exact same `my_function`.

For types (enums, struct, unions, distinct) the namespace is not required unless the name is ambiguous. Note also that the current module always take precedence. So if you have a `Foo` in your current module, and you import a module with `Foo`, you can use `Foo` without disambiguating it with a namespace, as the local `Foo` will always be assumed.

Even if you use a lot of modules, you should in general not need anything other than the required namespaces, that is the shortened namespace on constant/global/function/macro and no namespace on types.

Also, note that you never place the namespace on declarations.

**Reply by LukyGuyLucky** [reply]
I am sorry but may I ask in whichever case the `shortened namespace` is required when making use of the symbols inside that module from other module,say the main?
module m1;
fn void test1(){}
module main;
import m1;
fn void main() { test1();} // Error: Functions from other modules must be prefixed with the module name.
fn void main() { m1::test1();} //ok
I am wondering whether the first call is ok if there is unambiguous function/method definition.

Regards,
Peter

**Reply by lerno** [reply]
No, the shortened namespace is mandatory. There *is* one way to escape this, but that is for functions that should be regarded as builtins, and should be used with caution:

```c
fn void test1() @builtin {}
```

In regards to methods, namespacing is not really applicable.

**Reply by LukyGuyLucky** [reply]
So just to take example with functions,say,when interop with raylib,the c3 source could be containing lots... of raylib:: Even in language complicated as C++,when 
#include <iostream>;
using namespace std;
one can write
cout<<"Hi" ,no need to write std::cout<<"Hi";

I was just worrying so many raylib::  or rl:: do not look good ,if at the meanwhile other 3rd party libraries introduced in.

Anyway,this is only my own taste.Thanks for the prompt response.
Regards,
Peter

**Reply by lerno** [reply]
The idea is that it's basically following C convention. So if there is a C function like this:

```c
extern int raylib_do_something(int x);
```
With usage
```c
int x = raylib_do_something(y);
```
Then the corresponding function / usage in C3 would be:
```c
extern fn int do_something(int y);
...
int x = raylib::do_something(y);
```

**Reply by LukyGuyLucky** [reply]
Thanks.This seems good.

**Reply by LukyGuyLucky** [reply]
I am sorry but wait a moment please--

  raylib_do_something(int y)  vs 
(raylib::)do_something(int y)
their signature are totally different.Would you please briefing what happened here?

**Reply by lerno** [reply, marked as answer]
If you would actually link against the C function called `raylib_do_something` you'd use `@extern(...)`, like this:

```c
extern fn int do_something(int y) @extern("raylib_do_something");
```

**Reply by LukyGuyLucky** [reply]
ah yes,I forgot we have @extern.

Appreicated.
